libvast/vast/option_declaration_set.hpp

```diff
@@ -0,0 +1,135 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "vast/optional.hpp"
+#include "vast/expected.hpp"
+#include "vast/data.hpp"
+
+#include "vast/detail/steady_map.hpp"
+
+namespace vast {
+
+class option_map;
+
+/// A set of `option_declarations` that can fill an `option_map` from a CLI
+/// string.
+class option_declaration_set {
+public:
+  /// Wraps the parse state.
+  enum class parse_state {
+    successful,
+    option_already_exists,
+    begin_is_not_an_option,
+    name_not_declartion,
+    arg_passed_but_not_declared,
+    arg_declared_but_not_passed,
+    failed_to_parse_argument,
+    type_not_parsebale,
+    mismatch_of_default_value_and_arugment,
+    in_progress
+  };
+
+  /// A declaration of a CLI argument option.
+  class option_declaration {
+  public:
+    /// Constructs a declation of an option.
+    /// @param long_name A long name that identifies this option.
+    /// @param short_names A vector of short name that identifies this option.
+    /// @param description A decprition to this option.
+    /// @param has_argument A flag that describes whether this option has an
+    /// argument.
```

Comment:
Please indent documentation of parameters like this to make it easier to read:

```cpp
/// @param has_argument A flag that describes whether this option has an
///                     argument.
```

(I comment only on this occurrence, there are more)